### PR TITLE
Basic engine work: implement #19

### LINKED ISSUE #4: Implement Pengu's ECS into the repo

### LINKED ISSUE #19: Add simple level loading
We finally decided on how we're going to structure the basic stuff.
1) Have a list of YML files. Each one represents a _layout_, which can be either a level, UI element, loading/title screen, etc.
2) The YML contains a list of entities, each with a list of components with certain properties.
3) There will be no transformer hierarchy in the ECS because there is no need for it.

Here's how a normal YML layout file would look:
```yml
- import: "data/layouts/another_thing.yml" # Loads another level, apart from this one
- entity: # Defines an entity
  transform: # Defines a transform component. Max one component type per entity
    pos: { x: 10, y: 20 } # Defines the "pos" property of the transform component. This should ideally be controlled by attributes in the component's member variables, but for now they can be hardcoded on the level parser
- entity: # Another entity
  # ...
```
Objects that can only be loaded once, such as the player, use `singleton` instead of `entity`, and have an `id` property:
```yml
- singleton:
  - id: "player"
  # add components here
```

The plan is:
- [x] Add a `rect` namespace that contains a `State` struct which has the registry inside. Also have a `get_state` function which returns the global state.
- [x] Add a test component & system that for example prints something to the console, and that can be added to YML layouts using the `test:` string (Hardcode this into the YML layout parser for now)
- [x] Add a `void load_layout(std::string_view path)` function inside of the State struct that adds all the entities declared in a layout file.
- [x] Add a test variable to the component made earlier, and make the parser recognize and set it correctly.
- [x] Have an internal map that lists all singletons loaded until now for preventing repeated singleton entities.
- [x] Add a `config.yml` file in `data/` that contains a single item, which is the first layout to load:
```yml
- initialLayout: "data/layouts/loadScreen.yml" # for example
```

I choose you, @tccppgg/rect-engine, to implement this. Also, linking the [YAML wikipedia page](https://en.wikipedia.org/wiki/YAML) because it contains good YML examples to get the feel of it.

### PR DESCRIPTION
**Pull request by NotAPenguin0**
Closes #19 and closes #4.

### CODE REVIEW

File: src/engine/private/yml_importer.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+				// First, find the ID and check if the singleton already exists. Only if it doesn't we can add it
+				std::string id = elem->second["id"].as<std::string>();
+				// Not found, we can add it
+				if (state.singletons.find(id) == state.singletons.end()) {
+					ecs::entity_t entity = state.ecs.create_entity();
+					parseSingleton(elem->second, entity);
+					// Add to the singleton map
+					state.singletons[id] = entity;
+				}
+			}
```
Comment (alexdevteam) [inline src/engine/private/yml_importer.cpp:83]: We also need `- import: path` functionality, should also accept relative paths

**Review by alexdevteam — CHANGES_REQUESTED** [review]

File: src/engine/private/yml_importer.cpp (hunk 1 above)
Comment (NotAPenguin0) [reply src/engine/private/yml_importer.cpp:83]: Added in 52d8500

**Review by alexdevteam — APPROVED** [review]
this will not check for recursive imports, but eh, it's ok

**Comment by NotAPenguin0** [comment]
Recursive imports as in file A importing B and B importing A? That should probably hard error/crash anyway so I don't think it's a big deal.

**Pull request merged**